docs/docs/04-standard-library/sim/resource.md

```diff
@@ -0,0 +1,636 @@
+  var counter: num;
+
+  new(ctx: sim.IResourceContext) {
+    // startup code
+    this.counter = 0;
+  }
+
+  pub onStop() {
+    // shutdown code
+  }
+
+  pub inc(n: num?): num {
+    let p = this.counter;
+    this.counter += (n ?? 1);
+    return p;
+  }
+
+  pub peek(): num {
+    return this.counter;
+  }
+}
+```
+
+Each `CounterBackend` object that's created will have its own state.
+It will be initialized by the simulator with its constructor (the `new()` method), and shut down using its `onStop` method.
+
+From the perspective of `CounterBackend`, its `inc` and `peek` methods are always called serially, even if requests are made concurrently from other resources.
+
+Next, we have to define a preflight class which represents the service's frontend:
+
+```js
+class Counter {
+  backend: sim.Resource;
+
+  new() {
+    // this is a "backend factory". it returns an inflight class that implements the
+    // resource.
+    let factory = inflight (ctx): sim.IResource => {
+      return new CounterBackend(ctx);
+    };
+
+    this.backend = new sim.Resource(factory);
+  }
+
+  pub inflight inc(n: num?): num {
+    let response = this.backend.call("inc", [Json (n ?? 1)]);
+    return num.fromJson(response);
+  }
+
+  pub inflight peek(): num {
+    let response = this.backend.call("peek");
+    return num.fromJson(response);
+  }
+}
+```
+
+The `Counter` class is a simple wrapper around the `CounterBackend` class. It provides a way to interact with the resource using the `inc` and `peek` methods.
+
+`sim.Resource` has an inflight method named `call` that takes the name of a method to call and an array of arguments, and returns the serialized response from the resource.
+The arguments are serialized and sent to the resource over HTTP, which will process the request and returns a serialized response.
+
+> Note: Since the `call` method makes networked requests, it's important that the resource implementation provided by the inflight class responds in a timely manner.
+> If the resource takes over 30 seconds to respond, `call` will throw an error, and the simulator will terminate the resource.
+
+The `Counter` class can be used like this:
+
+```js
+let c = new Counter();
+
+new cloud.Function(inflight () => {
+  log("counter is {c.peek()}");
+  c.inc();
+  log("counter is now {c.peek()}");
+});
+```
+
+### Serializability
+
+Methods that are called on a `sim.Resource` instance must be serializable, meaning
+all of their arguments and return values must be serializable.
+
+The following types are serializable:
+
+- `num`
+- `str`
+- `bool`
+- `Json`
+- `Array`
+- `Map`
+- `T?` where `T` is a serializable type
+- enums
+- structs that contain only serializable types
+
+The `call` function on `sim.Resource` only supports arguments that are valid `Json` values.
+To convert a value to `Json`, use the `Json` constructor:
+
+```js
+let x: num = 5;
+let y: str = "a";
+this.backend.call("myMethod", [Json x, Json y]);
```

Comment:
```
let x: num = 5;
let y: str = "a";
let z = [x, y];
```

This is invalid in Wing since array elements have to be a single type. For the compiler to know you want to create an array of Json, you'd have to either make the elements `Json` or specify the array as Json.

```
let z = [Json x, Json y];
let z = Json [x, y];
```